micropython/micropython-lib
# usb: Add high-level USB device support packages

High-level support packages for creating USB device functionality in Python code.

These modules rely on the machine.USBDevice() object newly merged for MicroPython 1.23 (https://github.com/micropython/micropython/pull/9497).

**IMPORTANT**: The latest version of this PR requires a low-level change in MicroPython's USBDevice that was merged in https://github.com/micropython/micropython/pull/14253 - nightly builds from before 18 April 2024 do not have this change, and will fail in unexpected ways!!

## How to test/use this PR

This PR requires a nightly build of Micropython 1.23-preview or later, on a port with machine.USBDevice support (currently samd or rp2).

Can either edit manifest.py to freeze usb-device packages into the firmware, or use the following command to install packages directly from this PR:

```
$ mpremote connect /dev/ttyUSB0 mip install --index https://projectgus.github.io/micropython-lib/mip/feature/usbd_python usb-device-NAME
```

See the README in this PR for a list of package names.

## Example Code

See the `usb/examples/device` subdirectory for some simple example programs.

These aren't installed with the package but can be run as (for example) `mpremote run mouse_example.py`.

(Note: because any USB serial connection will close on reenumeration, printed output from this point will only be visible if outputting stdout to traditional UART, semihosting, etc.)

## Debugging Tips

See the micropython PR https://github.com/micropython/micropython/pull/9497

## TODOs

This PR is nearly complete, aside from code review comments I have a few things left to finish:

* [ ] Complete README with some class reference information.
* [x] Re-test on Windows.
* [x] Test all examples on macOS (help wanted as I don't have a Mac on hand).

## Future work (not this PR)

* Add an async USB interface helper. As USB is host-initiated and mostly uses callbacks, there isn't a whole lot more work needed here - however some classes would benefit from async variants.
* See if we can provide a host helper program for generating HID Report Descriptors (or link to one that already exists). I tried implementing this in MicroPython so it builds the Report Descriptor on the device at runtime, but this takes up too much space.

*This work was funded through GitHub Sponsors.*
